### ISSUE TITLE
optim.Optimizer should copy "params" before modifying them

### ISSUE THREAD

**Issue by AnthonyBarbier** [pytorch/pytorch#51794]
## To Reproduce

```py
import torch

model = torch.nn.Sequential(torch.nn.Linear(10, 10),torch.nn.Linear(10, 10))
params = [{"params":model[0].parameters()},{"params":model[1].parameters()}]
print(f"Params before {params}")
a = torch.optim.SGD(params, lr=0.01)
print(f"Params after {params}")
b = torch.optim.SGD(params, lr=0.5)
print(f"Expected 0.5, got: {b.param_groups[0]['lr']}")
```

Trace:
```
Params before [{'params': <generator object Module.parameters at 0x7f3a4c791678>}, {'params': <generator object Module.parameters at 0x7f381c6ee1a8>}]
Params after [{'params': [Parameter containing:
tensor([[...]], requires_grad=True), Parameter containing:
tensor([...], requires_grad=True)], 'lr': 0.01, 'momentum': 0, 'dampening': 0, 'weight_decay': 0, 'nesterov': False}, {'params': [Parameter containing:
tensor([[ ...]], requires_grad=True), Parameter containing:
tensor([...], requires_grad=True)], 'lr': 0.01, 'momentum': 0, 'dampening': 0, 'weight_decay': 0, 'nesterov': False}]
Expected 0.5, got: 0.01
```

## Expected behavior

params shouldn't be modified behind the user's back: torch should make a copy of it before using it to store the optimizer's state in it.

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip

**Comment by vincentqb** [comment]
Here's what I see:
```python
In [1]: import torch

In [2]: torch.__version__
Out[2]: '1.7.1'

In [3]: model = torch.nn.Sequential(torch.nn.Linear(10, 10),torch.nn.Linear(10, 10))

In [4]: params = [{"params":model[0].parameters()},{"params":model[1].parameters()}]

In [5]: params
Out[5]: 
[{'params': <generator object Module.parameters at 0x7f87b01a6580>},
 {'params': <generator object Module.parameters at 0x7f87b01a6430>}]

In [6]: a = torch.optim.SGD(params, lr=0.01)

In [7]: params
Out[7]: 
[{'params': [Parameter containing:
   tensor([[-0.2736,  0.1357,  0.1041, -0.2415, -0.0619,  0.2792,  0.2769,  0.1774,
            -0.1615, -0.2426],
           [ 0.2549,  0.1987, -0.1161, -0.2579,  0.1597,  0.2528,  0.0474, -0.2789,
            -0.1462,  0.2206],
           [-0.2796, -0.1662,  0.1666,  0.1996, -0.2504,  0.1458,  0.1916,  0.2521,
            -0.2762,  0.0212],
           [ 0.1446, -0.2476, -0.1077, -0.0343, -0.0968,  0.0697, -0.0147,  0.0828,
             0.2368,  0.0945],
           [ 0.0428,  0.3064,  0.1992, -0.1489,  0.1924, -0.2704,  0.1063,  0.0883,
             0.1947, -0.1834],
           [ 0.0709,  0.0274,  0.1337, -0.1455,  0.0280, -0.0018, -0.0872, -0.2851,
            -0.0417, -0.1944],
           [ 0.2133, -0.2722, -0.2057,  0.1187,  0.2016, -0.0028,  0.2229, -0.1767,
            -0.1576,  0.2023],
           [ 0.2996, -0.0221, -0.1570, -0.2026,  0.2746, -0.0164,  0.2859,  0.3085,
             0.1094,  0.2101],
           [-0.0887,  0.0866, -0.2534,  0.2306, -0.2811, -0.2330,  0.0314,  0.0224,
            -0.1074, -0.0180],
           [-0.2325,  0.1007, -0.1570,  0.2159, -0.2015, -0.0879, -0.0730,  0.0111,
             0.1482,  0.1912]], requires_grad=True),
   Parameter containing:
   tensor([-0.1213,  0.1143,  0.1274, -0.1482,  0.0317, -0.1760, -0.2197,  0.1506,
           -0.0872, -0.2227], requires_grad=True)],
  'lr': 0.01,
  'momentum': 0,
  'dampening': 0,
  'weight_decay': 0,
  'nesterov': False},
 {'params': [Parameter containing:
   tensor([[ 0.2370, -0.0706,  0.2927, -0.2290,  0.0421,  0.0443,  0.0170, -0.0355,
            -0.2966,  0.1205],
           [-0.1208,  0.2687, -0.1640,  0.0007, -0.0655, -0.1556,  0.2417,  0.2340,
             0.2729,  0.2876],
           [ 0.0798,  0.2402, -0.2152,  0.2171, -0.0854, -0.3160,  0.2199,  0.1523,
            -0.2555, -0.1886],
           [-0.2250,  0.2191,  0.2602,  0.3114,  0.1972,  0.0613, -0.1719,  0.2647,
             0.0938,  0.0333],
           [ 0.2193,  0.2078, -0.2975,  0.3075, -0.1322, -0.2935, -0.2242, -0.1055,
             0.1343,  0.0673],
           [ 0.2232,  0.0592,  0.2585, -0.3083,  0.2973,  0.1522,  0.0777, -0.2854,
            -0.1043, -0.0983],
           [-0.2181,  0.1842,  0.0327,  0.2822,  0.2450,  0.0314,  0.1939,  0.3022,
             0.1175,  0.2602],
           [ 0.3021, -0.0131, -0.2872, -0.2701, -0.1430, -0.2581,  0.0580, -0.3025,
             0.0871,  0.2136],
           [-0.3114, -0.0949,  0.0431, -0.1024,  0.0894, -0.1514, -0.2275,  0.2798,
            -0.0577, -0.1996],
           [ 0.3127, -0.0407, -0.2713,  0.1745, -0.1826, -0.1157, -0.1326, -0.1400,
            -0.2755,  0.2493]], requires_grad=True),
   Parameter containing:
   tensor([-0.0695,  0.2418,  0.1434,  0.1102,  0.0923,  0.2288, -0.1691, -0.2833,
            0.0818,  0.0664], requires_grad=True)],
  'lr': 0.01,
  'momentum': 0,
  'dampening': 0,
  'weight_decay': 0,
  'nesterov': False}]

In [8]: b = torch.optim.SGD(params, lr=0.5)

In [9]: params
Out[9]: 
[{'params': [Parameter containing:
   tensor([[-0.2736,  0.1357,  0.1041, -0.2415, -0.0619,  0.2792,  0.2769,  0.1774,
            -0.1615, -0.2426],
           [ 0.2549,  0.1987, -0.1161, -0.2579,  0.1597,  0.2528,  0.0474, -0.2789,
            -0.1462,  0.2206],
           [-0.2796, -0.1662,  0.1666,  0.1996, -0.2504,  0.1458,  0.1916,  0.2521,
            -0.2762,  0.0212],
           [ 0.1446, -0.2476, -0.1077, -0.0343, -0.0968,  0.0697, -0.0147,  0.0828,
             0.2368,  0.0945],
           [ 0.0428,  0.3064,  0.1992, -0.1489,  0.1924, -0.2704,  0.1063,  0.0883,
             0.1947, -0.1834],
           [ 0.0709,  0.0274,  0.1337, -0.1455,  0.0280, -0.0018, -0.0872, -0.2851,
            -0.0417, -0.1944],
           [ 0.2133, -0.2722, -0.2057,  0.1187,  0.2016, -0.0028,  0.2229, -0.1767,
            -0.1576,  0.2023],
           [ 0.2996, -0.0221, -0.1570, -0.2026,  0.2746, -0.0164,  0.2859,  0.3085,
             0.1094,  0.2101],
           [-0.0887,  0.0866, -0.2534,  0.2306, -0.2811, -0.2330,  0.0314,  0.0224,
            -0.1074, -0.0180],
           [-0.2325,  0.1007, -0.1570,  0.2159, -0.2015, -0.0879, -0.0730,  0.0111,
             0.1482,  0.1912]], requires_grad=True),
   Parameter containing:
   tensor([-0.1213,  0.1143,  0.1274, -0.1482,  0.0317, -0.1760, -0.2197,  0.1506,
           -0.0872, -0.2227], requires_grad=True)],
  'lr': 0.01,
  'momentum': 0,
  'dampening': 0,
  'weight_decay': 0,
  'nesterov': False},
 {'params': [Parameter containing:
   tensor([[ 0.2370, -0.0706,  0.2927, -0.2290,  0.0421,  0.0443,  0.0170, -0.0355,
            -0.2966,  0.1205],
           [-0.1208,  0.2687, -0.1640,  0.0007, -0.0655, -0.1556,  0.2417,  0.2340,
             0.2729,  0.2876],
           [ 0.0798,  0.2402, -0.2152,  0.2171, -0.0854, -0.3160,  0.2199,  0.1523,
            -0.2555, -0.1886],
           [-0.2250,  0.2191,  0.2602,  0.3114,  0.1972,  0.0613, -0.1719,  0.2647,
             0.0938,  0.0333],
           [ 0.2193,  0.2078, -0.2975,  0.3075, -0.1322, -0.2935, -0.2242, -0.1055,
             0.1343,  0.0673],
           [ 0.2232,  0.0592,  0.2585, -0.3083,  0.2973,  0.1522,  0.0777, -0.2854,
            -0.1043, -0.0983],
           [-0.2181,  0.1842,  0.0327,  0.2822,  0.2450,  0.0314,  0.1939,  0.3022,
             0.1175,  0.2602],
           [ 0.3021, -0.0131, -0.2872, -0.2701, -0.1430, -0.2581,  0.0580, -0.3025,
             0.0871,  0.2136],
           [-0.3114, -0.0949,  0.0431, -0.1024,  0.0894, -0.1514, -0.2275,  0.2798,
            -0.0577, -0.1996],
           [ 0.3127, -0.0407, -0.2713,  0.1745, -0.1826, -0.1157, -0.1326, -0.1400,
            -0.2755,  0.2493]], requires_grad=True),
   Parameter containing:
   tensor([-0.0695,  0.2418,  0.1434,  0.1102,  0.0923,  0.2288, -0.1691, -0.2833,
            0.0818,  0.0664], requires_grad=True)],
  'lr': 0.01,
  'momentum': 0,
  'dampening': 0,
  'weight_decay': 0,
  'nesterov': False}]
```

The parameters are materialized when defining the first optimizer, and the learning rate is attached to the model parameters as expected. The parameters (including the learning rate) are not changed when defining optimizer `b` (as can be seen [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L264)).

I'm assuming you were instead expecting the optimizer to update the learning rate in the model parameters following the definition of b. Is that the case?

> params shouldn't be modified behind the user's back: torch should make a copy of it before using it to store the optimizer's state in it.

By design, the learning rate is attached to the model parameters. Since the optimizers' goal is to update the model parameters, the optimizer is not making a copy, and modify the parameters directly. If you want to modify a copy, you can do so before passing the parameters to an optimizer. 

Is that what you meant?

**Comment by AnthonyBarbier** [comment]
Not really, I'm not contesting the changes to the model's parameters.

The problem I've got is with the `params` variable in the example case: after the first Optimizer has been instantiated the `params` variable ends up containing the state of that optimizer.

In the trace you pasted above it's the `out[7]`: 
```
  'lr': 0.01,
  'momentum': 0,
  'dampening': 0,
  'weight_decay': 0,
  'nesterov': False}]
```

And as a result when the second Optimizer is instantiated they end up sharing the same states because somehow the state of the optimizer was added to this input parameter.

As I said, I think all you need to do is to make a copy of params before you start adding extra states to it.

**Comment by vincentqb** [comment]
> And as a result when the second Optimizer is instantiated they end up sharing the same states because somehow the state of the optimizer was added to this input parameter.

By design, the state of the optimizer is attached to `params`. For instance, sgd reads `weight_decay`, `momentum`, `dampening`, `nesterov`, and `lr` from the params, see [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L89). Is that what you refer to? Since this behavior exists since the beginning of the optimizers, this would be difficult to change. However, I do agree that this could be added to the [documentation](https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L19). Would that help?
